### ISSUE TITLE
Mutators distributed with Blade of Agony

### ISSUE THREAD

**Issue by Username-N00b-is-not-available** [Realm667/WolfenDoom#459]
Starting a separate issue so as not to derail #297 further. This issue is the place to discuss the separate mutators @Tormentor667 said about in https://github.com/Realm667/WolfenDoom/issues/297#issuecomment-812902744.
I had some ideas on the addons we could ship with BoA, namely:

[**Bold** = done/mostly done, regular = in progress, _italic_ = not started]

**(1)** A melee weapon-only combat mode, where the player can only use knives, shovel, and the Firebrand (aside from the quick kick). This is done by replacing all weapon/item actors via ```CheckReplacement``` in an EventHandler with an empty dummy ```Nothing``` actor (**question**: actors which do not spawn on the map, for example, those which are given by scripts, are not subject to such 'removal'; @AFADoomer, @Talon1024, do you by chance know of a good way to deal with them?)
As a compensation, the player always has berserk damage, and the knife has an alternative ranged throwing attack (not sure if it is a good idea to throw combat knives, but we can change the sprites). You should be able to pick the knife up again after it killed an enemy. Additional knives should likely replace large ammo pickups or weapons with a small probability.
I think that @AFADoomer's dragging feature should also be turned on for this.
**(2)** Additional weapons (or alternative ones, since we already have a large arsenal). For example, I have made a scoped FG42 out of old ww-nazis resources and @DoomJedi's new scope. Maybe we could add an StG44 (not sure what to do with the ammo type), or a hunting shotgun, or dual pistols, or even the Snooper rifle directly from RTCW :-D Though there is a problem of drawing good enough sprites for most of these.
(3) ```InstantReaction``` MAPINFO keyword and other enemy behaviour changes.
_(4)_ Movement/game physics quirks, just like in the Unreal series.
**(5)** Smart Scavenger adaptation for Blade of Agony pickups.
**(6)** Instagib (everything is killed in one hit, including the player).
**(7)** Vampire (after 30 seconds of not killing anyone you start taking damage which increases over time).
**(8)** Fliegerfaust replacement on the truck sequence (M2HB with HE rounds %-) ).

**Current status** (09 Apr 2021, morning):
<details><summary>List (with videos)</summary>

(1): **done**, suggestions? [melee.zip](https://github.com/Realm667/WolfenDoom/files/6273502/melee.zip)

https://user-images.githubusercontent.com/59409713/113919531-60615680-97ec-11eb-8ae5-7d0afac37d84.mp4

(2): **mostly done**, needs two more Kampfpistole frames and StG 44 magazine sprites. [more_weapons.zip](https://github.com/Realm667/WolfenDoom/files/6265810/more_weapons.zip)

Features StG 44 (slot 4), Kampfpistole (slot 6), and FG 42 (slot 7).
- StG 44 is dropped only by enemies who wield it (WaffenSS and derivatives) and has a separate ammo type. Has high DPS and higher damage per bullet than the SMGs have, the roles are divided just like with MP 43 and MP 40 in Wolf2009.
- Kampfpistole uses Nebelwerfer rockets. Basically it is an ersatz Nebelwerfer which also shoots impact grenades instead of rockets and has around 2x lower DPS. The weapon is dropped from some Wehrmacht and SS officers, so it could even appear in C1 (these types of officer are spawned with chance 1 in 8 and can be distinguished by players by their longer aiming delays). C1M0 has supermutants who drop rocket crates, just saying ;-)
- FG 42 can only be randomly obtained from some of the paratroopers. Quite OP because of high damage per shot, full auto, high DPS and common ammo type. (The scope is also functional, of course.) It has lowered fire rate and increased recoil, so firing very long bursts might not be efficient.

https://user-images.githubusercontent.com/59409713/113919764-a28a9800-97ec-11eb-94b2-ffffa4b9f6c8.mp4

(3): Leaping mutants are done. Add ```InstantReaction```, target leading snipers and crouch firing soldiers. [behaviour.zip](https://github.com/Realm667/WolfenDoom/files/6281106/behaviour.zip)

https://user-images.githubusercontent.com/59409713/114086977-fd3ef500-98bb-11eb-8997-e787c909963c.mp4

(4): _not started yet_.
(5): **mostly done**, based on Thrifty Ammo/Health by zrrion the insect. [efficiency.zip](https://github.com/Realm667/WolfenDoom/files/6278796/efficiency.zip) The new behaviour is as such: if the new ammo/health item fits into your inventory completely, you just pick it up automatically when stepping over it, as usual. If the item does not fully fit (e. g. it is a large 40-health pack, and you have 63 health and no health upgrades), then you won't pick it up by stepping over it, but you are still able to take it by pressing the use key. It gives the player an option to decide on picking the item up themselves. Works for all major ammo/health pickups.

https://user-images.githubusercontent.com/59409713/114030634-458df100-9883-11eb-8ad0-3dd963deaa1d.mp4

(6): **done**. [instagib.zip](https://github.com/Realm667/WolfenDoom/files/6272966/instagib.zip)

https://user-images.githubusercontent.com/59409713/113918995-be416e80-97eb-11eb-8f9a-6f6438d1f759.mp4

(7): **mostly done**, thinking on how to make better use of dogs and Werewaffen-SS. Also giving blood when the player kills enemies by proxy (e. g. with barrels) needs to be tested and debugged. [vampire.zip](https://github.com/Realm667/WolfenDoom/files/6284738/vampire.zip)

An enemy kill gives you additional 5 seconds. You can stack up to 30 seconds. Once the time runs out, you start taking damage (which linearly increases over time, drowning-style), once every 64 tics. The enemy kill which ends your blood thirst gives 2x additional time (=10 seconds). At the start of every map there is also a ten-second handicap (this is for e. g. cutscenes).

(8): **done** (the weapon was quite easy, the problem was to replace it in the scripted sequence properly from ZScript, i. e. without editing the map/ACS). [m2hb.zip](https://github.com/Realm667/WolfenDoom/files/6272971/m2hb.zip)

https://user-images.githubusercontent.com/59409713/113919258-111b2600-97ec-11eb-93bb-685eee7a8232.mp4

</details>

Does anyone have comments, or, better yet, other addon ideas? Please write about them here.

**Comment by DoomJedi** [comment]
This is a bit too much - quite a bit too late to add to a well tested mod near the release. Those would not only be rushed, but could introduce many potential bugs to the actual game.
I don't like this direction of thought for reason mentioned, we should focus more on testing and not adding new mechanics, which can be done in later update.

Not to mention I develop my own mods these days, Passover holidays are over, I have no experience in making Doom weapons art (scope was my first little Doom-style weapon edit ever).

Now, if I totally ignore the above said for a moment and will refer to the actual suggestions - there are alot of interesting and easy to use mutators we can use, my SplitWolf mod has many single-player game modes I invented or remade for single-player.
From obvious and stressful Instagib, to Bomb Defuse (bomb sprite is randomly placed in a map and you have limited time (related to par time multiples) to reach it and defuse, Vampire mode (you health decrease if you don't kill new enemies frequently enough to feed off their spawn hearts), Harvester and other mods you can see and play in SplitWolf.
Splitwolf also has a "melee Berserk Mod" similar to the one Noob described.
I can sure brainstorm more idea.

Sure it's a place to remind you of the unused leaping mutants that would make BoA more challenging.

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi
> could introduce many potential bugs to the actual game

The mutators will not be included in BoA, but only _distributed_ along with it as optional addons which are turned off by default. They will not affect the base game at all. As for being rushed, I agree, but these are not supposed to be complex several-MB gameplay mods, just small 'patches' (hence the 'mutators' designation) that have a small but game-changing feature.

The SplitWolf gamemodes you mention are definitely something, they also reminded me of some mod ideas I had a few years ago, thank you for the memories :-) I'll research a bit more on SplitWolf.

> I can sure brainstorm more idea.

This is also quite encouraging, please do it :-)

Yes, leaping mutants would fit nicely, but BoA uses the mutants so rarely this change _alone_ would not make much of a difference. I can only remember C1M2, C1M0, C2M1, C2M3, C2M5, and C2M6_A, and the numbers of melee/flying hack mutants are even smaller. Though I should probably talk less and do more, then we'll see how it turns out...

**Comment by DoomJedi** [comment]
Recent "fresh" SplitWolf gameplay video from March 22, they try quite some modes there (the hud bug is not from the mod, probably they used some wrong file versions)
https://www.youtube.com/watch?v=pn8xQ6AioWM

My "Team Raycast" has a different attitude - not to allow different way to play the mod before people get to explore the mod and its maps in its intended gameplay way (and even spoilery bonus maps were locked (as you know by now) to be allowed only after all or most of the mod is explored enough). That's why none of TR cheats is publicly known - still - for any of our released mods.
Not that this is the best way or approach, just saying...when you plan a mod you plan certain experience to be had before you make twists on in.

**Comment by Username-N00b-is-not-available** [comment]
Updated the starting post with some intermediate progress on the mutators, in case anyone else is interested in them.

**Comment by Tormentor667** [comment]
I like where this is leading to ;)

**Comment by DoomJedi** [comment]
Yes, there are plenty of weapons in Wolfendoom mods and such, no need to make new art.
Just to make sure the hand/glove is consistent.

**Comment by Username-N00b-is-not-available** [comment]
Thank you!
I wish it was so easy. Hands are not the only problem: most of the suggested weapon sprites are done in a style completely clashing with BoA weapons. This is also true of ww-nazis weapons, since BoA arsenal has been updated many times during the 6 years of development, but it stands out even more for Las Rojas ones. I don't believe that you would let sprites of this kind anywhere near Blade of Agony:
![image](https://user-images.githubusercontent.com/59409713/113612072-06c82300-9658-11eb-84d1-6af2754589d2.png) ![image](https://user-images.githubusercontent.com/59409713/113611082-b3a1a080-9656-11eb-8262-6277b641acfe.png) ![image](https://user-images.githubusercontent.com/59409713/113611118-bef4cc00-9656-11eb-9428-aa87b914c76f.png)
There's also a question of weapon placement. StG44s and FG42s could be dropped by Waffen-SS and Paratroopers (yes I know about the sprites, let's not look at it too seriously :-D ), but I don't see where you would get Allied weapons and especially ammo for them on the levels.

It seems I have found some better sprites for the StG44, a bit later I'll try editing and Frankensteining them and see if they could make a good addition:
![43a3Uua](https://user-images.githubusercontent.com/59409713/113619854-2cf2c080-9662-11eb-9824-48d39ecdd349.png) ![s5e7heY](https://user-images.githubusercontent.com/59409713/113619855-2d8b5700-9662-11eb-8fd0-25722125860f.png)

**Comment by DoomJedi** [comment]
I want to make a few things as clear as possible.

I had nothing to do with BoA weapons (HUD weapon sprite-wise).
I have no skills artwise to make new weapons, this is a particular expertise I don't have even in the slightest.
So if you plan on any new weapons - count me out and find a person who is up for the task.

I also think that to add mutators at the last minute that require significant new art is not right.
I also think that mutators - can be later update, after enough people played enough of BoA as intended by its makers. I think the focusing on them now (and moreso rushing them), near release, is not right set of priorities and we should focus more on testing.

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi The first part of your message has been all clear to me for quite a while. I have not requested you to make any weapon sprites, and will not do it, I know you would have refused. I am doing the sprites myself right now.
As for the last paragraph, I can agree only partly. These suggestions would be alright if the BoA team stayed intact after the release. Tormentor667 is retiring, and keeping in mind the overall activity level we have on the repository now, in the most important time before the release, can I really expect anyone from the team to quality control them _after_ it? We sure can add some of the mutators as additional downloadable content after the release, but I think I must finish them ASAP.
I think this type of activity is the best what I could myself offer the team before the release, since it is not possible for me to test current commits, and even if it was, one full testing round would take a whole lot of time from me, way more than even doing the art.
In the end, if some of the graphics and addons I am currently doing turn out to be too subpar, we can just throw the shitty parts out and call it a day.

**Comment by DoomJedi** [comment]
As long as I'm not expected to do those - naturally it's less important for me if it'll be done or not, and the idea itself is cool (and I support it) and one I can even help brainstorming for. I sure wish you best luck with the new weapons, such skills are very useful in Doom modding in general even afterwards.

I see your point, and Tormentor leaving indeed put priorities in different light (though I remember he said he will support BoA afterwards?) and the most we can do with him onboard - the better. And sure there is larger hype for this now from others as well, to test and help, to support.

**Comment by Username-N00b-is-not-available** [comment]
I seem to have run out of ideas on the weapons, so here is a pre-release version of the weapon addon: [more_weapons.zip](https://github.com/Realm667/WolfenDoom/files/6265810/more_weapons.zip)
It features StG 44 (slot 4), Kampfpistole (slot 6), and FG 42 (slot 7).
- StG 44 is dropped only by enemies who wield it (WaffenSS and derivatives) and has a separate ammo type. Has high DPS and higher damage per bullet than the SMGs have, the roles are divided just like with MP 43 and MP 40 in Wolf2009.
- Kampfpistole uses Nebelwerfer rockets. Basically it is an ersatz Nebelwerfer which also shoots impact grenades instead of rockets and has around 2x lower DPS. The weapon is dropped from some Wehrmacht and SS officers, so it could even appear in C1 (these types of officer are spawned with chance 1 in 8 and can be distinguished by players by their longer aiming delays). C1M0 has supermutants who drop rocket crates, just saying ;-)
- FG 42 can only be randomly obtained from some of the paratroopers. Quite OP because of high damage per shot, full auto, high DPS and common ammo type. (The scope is also functional, of course.) It has lowered fire rate and increased recoil, so firing very long bursts might not be efficient.

Known problems: too few frames for Kampfpistole reloading; no magazine sprites for StG 44 ammo.

**Comment by DoomJedi** [comment]
This reads like a great addition to the weapon arsenal.

**Comment by Username-N00b-is-not-available** [comment]
Five addons have been done, and I've made some short video clips that showcase them (see first post).
@Tormentor667, what are your opinions on these addons? What should be changed or added?

**Comment by Tormentor667** [comment]
I love what you are doing here @Username-N00b-is-not-available , I really love it :) I am fine with those who are finished by now!

I just wonder (this goes for @madame-rachelle though) how we can add some kind of "tooltip" or "explanation" to the addon launcher, so people know what each mutator/addon does. By the way, also for Rachelle, do you think you can prepare the changes soon enough so we can test them and finish everything for the release on April 30th?

**Comment by DoomJedi** [comment]
Looks great so far pal, adds much of replay value to BoA.

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi Adding the leap mutant right now, seems almost done :-)

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi 

https://user-images.githubusercontent.com/59409713/114086977-fd3ef500-98bb-11eb-8997-e787c909963c.mp4

**Comment by Username-N00b-is-not-available** [comment]
Now that most of the 8 mutators are in quite a decent state by my standards, does anyone have ideas for (3) and (4) from my original post?
(3) is for changes in the default enemies, such as seen in the posted mutant video. The mentioned sniper target leading has also been implemented. I tried to think of more changes that would make their behaviour more complex, but ran into a problem...
As for (4), I'm completely out of ideas here. It could be better to scrap it, but we can possibly think of something.
By the way, any other new addon topics would also be greatly appreciated ;-)
***
Also, does anyone [except myself, of course] want to test some of these and post feedback here? @x3003, would you like to do it?
The starting post has links to the archives under the spoiler. As for now the launcher addon system has not been implemented yet, but you can just drag the .boa file onto the GZDoom executable, it'll work just fine.

**Comment by DoomJedi** [comment]
Ok, time for brainstorming, DoomJedi style :) Gotta help my pal Noob :)

3. We need changes that do not require alot of new art, or use scrapped art - and that is the hardest part of the task. Because we don't have time, otherwise I would really have alot of ideas from my mods.
I have ideas but all have downsides...but maybe those are okay for mutators?
3.1 More enemies can go into a grenade throwing state, and not just special actor. This uses same art with little detail inconsistency.
3.2 All rifle enemies can become accurate snipers instead normal shooting. But they take more time to first shoot as they aim more accurately.
3.3 Nazis soldiers can go to sleep with high chance on night levels or long guarding without player engagement, not just special sleep actor. If outdoors that can even go to timed peeing state sometimes after long time of lack of engagement.
3.4 More nazi surrender scenarios.
3.5. Nazis can become kamikaze using grenade
3.6 Player can kick the grenade thrown at him back to the enemy, play soccer with it. 
3.7 AI - smart enemies - can retreat or hide if low on health and ammo, other such smart mechanics.
3.8 Advancements to the medic - how effectively he can heal others. Maybe "adrenaline" grenade that heals all enemies in the area at once once thrown
3.9 No mine detector?
3.10 Grenade rolling sometimes from dead enemy and explodes.
3.11 Heatseek rockets from enemies?

4. 
4.1 Low gravity? Can make maps played not as intended.
4.2 Mode where the more weapons you carry the slower you move from the weight and you need to drop unneeded weapons?
4.3 Mode where you can only carry 3 weapons at once (or only 1?), and have to choose what to carry?
4.4 Grenade only mode, you have unlimited grenades but no other weapons? Or a mod where you have endless ammo but only for one weaker weapon?
4.5 Ability to roll? Dash to close distance to the enemy? 
4.6 Auto-activated inventory once picked up?
4.7 Ability to throw manually activated grenades? Proximity grenades? 
4.8 Reloading throws away current clip with all bullets currently in it? Will make reloading more tactic.

**Comment by Username-N00b-is-not-available** [comment]
Great ideas, thank you very much, @DoomJedi! Here are some initial comments on them:
<details><summary>Comments</summary>

3.1 More enemies can go into a grenade throwing state, and not just special actor. This uses same art with little detail inconsistency.
> Yeah, at least the ones who drop the grenades in BoA (MP40 guards and Waffen). IIRC the elite soldiers do not use any separate sprites for throwing grenades, so this probably won't even clash with the base game too much.

3.2 All rifle enemies can become accurate snipers instead normal shooting. But they take more time to first shoot as they aim more accurately.
> We can also increase MaxTargetRange for them, 768 is ridiculously small.

3.3 Nazis soldiers can go to sleep with high chance on night levels or long guarding without player engagement, not just special sleep actor. If outdoors that can even go to timed peeing state sometimes after long time of lack of engagement.
> Would need careful and difficult balance, or we could have officers peeing on floor in their own cabinets :-)

3.4 More nazi surrender scenarios.
> Including false surrenderings where the guard gives away the primary weapon (rifle/mp40) and hastily shoots the player with pistol. Will not require too much art, especially if we only use Wehrmacht guards.

3.5. Nazis can become kamikaze using grenade
> Very nice idea for half-dead guards!

3.6 Player can kick the grenade thrown at him back to the enemy, play soccer with it.
> A stock BoA grenade lives less than a second on the ground. This will need overall rebalancing of grenade timing/range/damage, but it is quite doable, and a good idea.

3.7 AI - smart enemies - can retreat or hide if low on health and ammo, other such smart mechanics.
> I'd really like to implement this properly, but I fear it would take way more time than we have left. Maybe a simpler solution would be possible, e. g. they get +FRIGHTENED and run away from player in case the health is very low. Connected to 3.4.

3.8 Advancements to the medic - how effectively he can heal others. Maybe "adrenaline" grenade that heals all enemies in the area at once once thrown
> Replace medics by evil Thule cultists?

3.9 No mine detector?
> But why? :-D

3.10 Grenade rolling sometimes from dead enemy and explodes.
> Connected to 3.5. Maybe some enemies would drop (by DropItem) live grenades instead of grenade pickups, it is possible ;-)

3.11 Heatseek rockets from enemies?
> Do you mean just homing rockets, like that of Doom revenants, or something extra? This could make some boss fights really frustrating, see the last keycard fight on C2M0_B with Tiger Roboters (their projectiles are homing).

3.12 Peeing adds health, you can also drink health from broken toilets (Duke style).
> This would add purpose to an integral and the most ubiquitous part of BoA maps :-)

4.1 Low gravity? Can make maps played not as intended.
> And make players stuck in some unintended places... Same with changing player dimensions.

4.2 Mode where the more weapons you carry the slower you move from the weight and you need to drop unneeded weapons?
> Realistic and quite possible to do (not counting inventory problems). Though to make this mode fun to play, the 'base' speed without weapons needs to be raised.

4.3 Mode where you can only carry 3 weapons at once (or only 1?), and have to choose what to carry?
> That would likely be a lot of code and even more debugging, considering the inventory system we have to work in. I like 4.2 way more.

4.4 Grenade only mode, you have unlimited grenades but no other weapons? Or a mod where you have endless ammo but only for one weaker weapon?
> Some years ago I made a Doom mod for myself where the only player weapon was a jammed minigun that just could not stop firing. The status bar also had a counter of how much money would have been spent on ammo :-) With unlimited grenades this could be great fun, much better than Serious Sam's revolvers with techno-magical ammunition replenishers. Just needs balance. Also there is already a melee-only mutator (but there are multiple weapons).

4.5 Ability to roll? Dash to close distance to the enemy?
> We already have the sprint mode, I would say that dash would just have duplicated it. As for rolling, it could be of use. We have screen tilting already in the mod. I don't want to do BD/PB-style separate keybinds, but I seem to know another way to trigger it...

4.6 Auto-activated inventory once picked up?
> Not sure what you mean here. This would ruin the whole idea of e. g. minesweepers in BoA, since they have use on levels that do not contain them as pickups.

4.7 Ability to throw manually activated grenades? Proximity grenades?
> Well, we have frag mines (in C3 you can buy them from shop). I never use them :-) , but maybe they could see more appearances in custom maps (that is beyond the scope of our mutators). I don't get the 'manually activated grenades' part, could you please elaborate?

4.8 Reloading throws away current clip with all bullets currently in it? Will make reloading more tactic.
> I had ideas for Red Orchestra-style reloading in Doom for quite a long time, but maybe such a simple solution is best.

</details>

I think I could definitely implement at least some of these. I just hope I have more time for them...

**Comment by AFADoomer** [comment]
A possible addition:
[Score.pk3](https://github.com/Realm667/WolfenDoom/files/6289315/Score.zip)

Adds a score counter with customizable point values for all enemies, arcade-style point amounts rising from dead bodies, and a coins/ammo/health bonus given every 40000 points, mirroring how Wolf3D gave extra lives for points.

**Comment by Tormentor667** [comment]
> A possible addition:
> [Score.pk3](https://github.com/Realm667/WolfenDoom/files/6289315/Score.zip)
> 
> Adds a score counter with customizable point values for all enemies, arcade-style point amounts rising from dead bodies, and a coins/ammo/health bonus given every 40000 points, mirroring how Wolf3D gave extra lives for points.

That one rocks - a must have in my eyes ;)

**Comment by DoomJedi** [comment]
Great ideas pal (Noob) :)

3.2 If you want to go even further, if player moved enough since last shot you can make the sniper to re-aim longer for sniper shot accuracy, which will motivate player to move more.

3.3 Well, for peeing you can check if it's outdoor like we did in Wolf3D mods...or in Doom there is no way to check for roofs/ceilings? Or you can go with just sleeping (some of those are my past suggestions for BoA, this one was originally declined because they wanted sleeping guards mainly in corners and in full mapper control, as it's 1-dir animation that might look less good otherwise, and I wasn't ready to make it 8dir).

3.5 Didn't think of it, but yeah, perfect for half-dead ones. Originally was thinking of rolling one too, but it's less realistic and they don't roll far enough to be safe.

3.7 Yes, frightened state is enough.

3.6 Another of my suggestions for BoA that didn't go in.

3.8 Good idea, some healing magic. But still, I think some adrenaline gas grenades can do the trick.

3.9 Never mind, forget it

3.11 You can do it only for RPG nazis.

3.12 From playthroughs most players try it anyway, and knowing Duke are surprised those don't work in BoA. Not only nice Easter Egg for a classic - but also nice way to balance the stronger enemies in this mutator.

4.1 Forget it then. Agree.

4.2 Yes, needs default speed rebalancing.

4.4 Can be some levels unpassable though, where you need to take down snipers to continue...just a remark.
Any weapon of big damage but very tactic use (with ability of self-damage in tight areas) can be made into a mutator, like flamegun-only mode or rpg-only mode.
Each such mutator will add really neat twist to BoA and replay value, whole game is different all of a sudden.

4.7 " I don't get the 'manually activated grenades' part," Meant Duke3D pipebombs.

If you need more ideas - can do more brainstorming, always bring up more. But you seem to have enough on your hands as is :)
Start with easier tasks and see how time you'll have left for harder ones.

**Comment by Username-N00b-is-not-available** [comment]
@AFADoomer This also shows n00b modders how to use your data parser. Very fine! Though it does not work properly with non-fullscreen HUDs, the score gets shown at upper right corner. I'll change it a bit.
@DoomJedi 4.4 No, it won't make completing such levels impossible ;-)

https://user-images.githubusercontent.com/59409713/114281158-7453c500-9a45-11eb-949d-0fe649818682.mp4

**Comment by DoomJedi** [comment]
I meant a different level, where you have to shoot 3 (?) snipers in the beginning of a mission from a cave.

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi Fair point on C2M5_A, the grenades will need a speed boost and/or a lower gravity coefficient.

**Comment by AFADoomer** [comment]
@Username-N00b-is-not-available - I just fixed a bug in one of the libraries that the Score add-on was using (55b98c7958d1a8b69efc6a281faf2680597a3fcf).  So display *should* be OK now for all screen sizes.

**Comment by Ozymandias81** [comment]
For BoA I always had 2 ideas that I never got to implement them, specially once ZScript changed everything for me... Things were the following:

1 - a grenade indicator that tells you where are you throwing the bomb, which updates as you hold the throw button (so physics should change this way... for an inventory item?)
2 - holy grenades, which makes the big spherical flash seen when Hitler dies at the end, in order to kill undead enemies and maybe distract human soldiers

**Comment by Username-N00b-is-not-available** [comment]
@AFADoomer, this happened when I killed a mechanic (he seemed to roll) with a flamethrower on C3M6_B:
![image](https://user-images.githubusercontent.com/59409713/114867065-028cca00-9dfd-11eb-9fd9-421c6dd5c151.png)

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi, what do you think of this? 

https://user-images.githubusercontent.com/59409713/114890015-51ddf500-9e13-11eb-877c-8d452d70db6e.mp4

**Comment by DoomJedi** [comment]
Hard to sleep with open eyes...why to use kneeing frames when we have sleeping frames for that guard?
Cause you want it to be generic?

**Comment by Username-N00b-is-not-available** [comment]
Honestly, I see no point in making only MP40 Wehrmacht guards able to sleep.
It may be possible to use frames from the surrendering soldier, but that also leaves us only with Wehrmacht guards...
Meanwhile I remade the grenade mod and am trying to implement what @Ozymandias81 suggested...

**Comment by DoomJedi** [comment]
Meantime recolored on DropBox afrika guard as well

**Comment by Username-N00b-is-not-available** [comment]
@DoomJedi We do not have such guards in the mod at all :-) The MP40 AK guard has a cap instead of a helmet.
I've recolored the necessary sprites myself.

https://user-images.githubusercontent.com/59409713/114933985-01ca5700-9e42-11eb-99ae-bb7c83cfe45e.mp4

As you can see, these drop their weapons to the ground when asleep. If a player steals the weapon, they draw pistols when woken up, but if not, they retake the weapon. Now I just need to think of a good way to check the light conditions for falling asleep.
***
And now I am also thinking of making pistol guards (AK, W, SS) able to pick up weapons for which we have frames (MP40, G43, shotgun). Plus my dreams about making grenadiers more dangerous...

**Comment by DoomJedi** [comment]
I have to say picking weapon from sleeping guards is a very neat idea, how I didn't think of that? :)

**Comment by Username-N00b-is-not-available** [comment]
https://user-images.githubusercontent.com/59409713/114985494-8b5a4300-9e9b-11eb-9c5c-af91278224e2.mp4

The AK guards do not pick up the shotgun and MP40 (I didn't allow them because they have no frames), but Wehrmacht guards are able to pick up any of three weapons. (To make things clear: I do not request drawing missing sprites here, just updating on my progress :-) )

I'm trying to wrap everything up with the mutators now, since I'm not sure if I have any more time in the following weeks...

**Comment by DoomJedi** [comment]
Dont forget toilets and other stuff :)

**Comment by madame-rachelle** [comment]
> 
> 
> I love what you are doing here @Username-N00b-is-not-available , I really love it :) I am fine with those who are finished by now!
> 
> I just wonder (this goes for @madame-rachelle though) how we can add some kind of "tooltip" or "explanation" to the addon launcher, so people know what each mutator/addon does. By the way, also for Rachelle, do you think you can prepare the changes soon enough so we can test them and finish everything for the release on April 30th?

We won't be able to have "details" or "tooltips" except what might be available to do via Windows Explorer. The loader interface is going to use a common shell dialog, pretty much nothing more.

**Comment by Username-N00b-is-not-available** [comment]
@madame-rachelle So, the mod loader interface won't be integrated into the launcher? Then there is no need to write a separate loader at all, since you can just drag and drop the .boa file onto GZDoom, which looks even easier than opening a command line script and writing something there.

**Comment by madame-rachelle** [comment]
No, the idea was something akin to "File->Open"

Nevertheless it's probably not going to work anyway - the upgrade from VS 2017 to VS 2019 destroyed the project for some reason. I don't know what's wrong or how it happened, and in order to fix it I may have to redo the whole thing from scratch. Ugh. This thing is just full of problems. :(

**Comment by Tormentor667** [comment]
This is not the kind of news I was hoping for 2 weeks before the release date. Sad to hear that, do you think you can somehow get it done?

Would it be possible to create a "dropdown" that automatically pulls addons from a /addon/ folder with files that have the .boa ending and the dropdown just shows the names of the .boa files? That way we could create a long filenime that explains what the mod is doing.

**Comment by madame-rachelle** [comment]
It's not possible to "just" do anything as long as I am unable to load the form for original project. I'll try and see what I can do but things aren't looking good. I am way out of my element with UI programming.

**Comment by Tormentor667** [comment]
Sorry for sounding like this is an easy task - I am just not used to how things are working :( I wonder, do you know someone who is able to support you with this? I could even get some bucks from the Realm667 financial pool if that helps in terms of getting an external programmer.

**Comment by madame-rachelle** [comment]
I don't, I wish I did, it would make this a lot easier.

I know someone who would be willing to take money but I doubt the R667 pool would be enough to persuade him to do it. He's does it for a living, isn't exactly poor, and hates UI programming, too. (Honestly, it sucks, because as easy as it is to design UI's on paper, actually making them work is a fucking nightmare)

**Comment by Tormentor667** [comment]
Could you ask him how much he'd like to have for that programming? Otherwise I'd try to open a thread on ZDoom, maybe we can find someone who is able to do that...

**Comment by Username-N00b-is-not-available** [comment]
I seem to have compiled an archive with mutators I had time to finish. 
In a nutshell, here is the archive that needs to be unpacked into the main BoA folder with gzdoom.exe: [addons.zip](https://github.com/Realm667/WolfenDoom/files/6330092/addons.zip). If it is not possible to implement the addon system into the launcher, a mutator could be loaded by dragging its header right onto the game executable.

**Comment by Tormentor667** [comment]
I moved the addons into my local "standalone" folder which I will prepare for the release.

**Comment by Tormentor667** [comment]
Reopened: How could we explain what each addon does?

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667 Maybe on the website + with an ```addons_readme.txt``` file in the archive... Though not sure if it is a good idea or not.

**Comment by Tormentor667** [comment]
Most probably the best choice. Can you prepare me the texts? I will put them online then :)

**Comment by Tormentor667** [comment]
Perfect, thanks a lot!

**Comment by Tormentor667** [comment]
I prepared something @Username-N00b-is-not-available , so if you can give me a little explanation for each addon (1-2 sentences, all about the same length) and one screenshot, I can turn it into a beautiful thing ;)

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667 How long the sentences should approximately be (in symbols)?

**Comment by Tormentor667** [comment]
About the same length as this:

> If you run into one of these with your personal Walther, intelligence suggests to run. No need to be very intelligent to agree. Make sure to have Panzerschreck with you, or even better: Your own fucking tank.

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667, here are the texts (general explanation + hints on each addon):

<details><summary></summary>

The mutators are small modifications made by the Blade of Agony team, but not included into the base game. These not only offer alternative ways to play Blade of Agony, but also give references to some features which can help future modders to know where they could start their work.

***

behaviour

Adds some advanced behaviour patterns to regular enemies, such as leaping for melee mutants, picking up weapons, sleeping, and throwing grenades at close range as an extreme measure after being wounded.

disable_hints
Removes all exclamation marks and secret hint sheets from the game, and disables the compass. This is for people who do not like in-game hints and think that Blade of Agony has a bit too much of them.

efficiency
Prevents wasting ammo and health items, disallowing to pick them up automatically if they do not fit fully in your inventory. If you still want to do it, press use on the item, and it will be picked up.

grenades_only
Out of all weapons, you can now only use grenades. The good news is that you can throw them further... and that you have an infinite supply of them. Don't get too careless in close quarters!

instagib
All damage from enemies to player and vice versa is multiplied by 1000. Terrain-based damage (e. g. from falling/drowning), as well as damage from critters, are excluded for the game to become playable.

m2hb
Replaces the Fliegerfaust mounted on the truck from C2M4 with a Browning .50 caliber M2HB heavy machine gun, loaded with high-explosive cartridges. It is a scourge to both enemy infantry and vehicles.

melee
Leaves only melee weapons in the game. Damage is buffed, and the knives can now be thrown (you can carry multiple ones). After a silent kill you can also drag the dead body away to hide it from patrols.

more_weapons
Adds three new weapons to the game. Kampfpistole is used by (and dropped from) some of the officers, StG 44 is a weapon of elite Waffen SS units, and FG 42 is sometimes used by paratroopers.

score
Upon killing an enemy, the player gains a few score points, just like in Wolfenstein 3-D. The amounts are larger for tough and rare enemies, and every 40000 points you receive some bonus goodies.

vampire
You are a vampire who quenches his thirst by killing Nazis and their accomplices. You've become so addicted to blood that you cannot wait even a minute without extending your track record.

</details>

@Talon1024 Could you please proofread these explanations before @Tormentor667 uses them? I think it's very likely that there are some phrases that could sound unnatural to native English speakers...

Screenshots: https://mega.nz/file/WqJCgB6L#l16pdY5TJI4wY81MmcJnTWmChymXgt-Rz8ZNZFt0oj0

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667, I spotted some problems with the addons (and I also forgot an explanation and a screenshot for one of the addons), I will update the archive and the previous post once I fix the issues.

**Comment by Tormentor667** [comment]
Already uploaded what's done so far :) Good work guys!

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667 I have not fixed the problems yet, you will have to upload this again...

**Comment by Tormentor667** [comment]
Do you mean the addons? No problem, I can do that. Can you just zip each of them as single file?
![2021-04-20_123333](https://user-images.githubusercontent.com/10298439/115382153-a6e38780-a1d4-11eb-820f-f438803d4dfb.png)
----------------------
By the way, I think it would be a good decision to release the addons about 2 weeks after the release of Chapter 3 - actually as a reason to "replay it". What do you think?

**Comment by Username-N00b-is-not-available** [comment]
> it would be a good decision to release the addons about 2 weeks after the release of Chapter 3 - actually as a reason to "replay it"

Yes, I agree with this, it would probably be better.

> Can you just zip each of them as single file?

OK, so I'll place there a .boa file, an ```addons/``` subfolder and put the main pk3 into the subfolder (so the player who downloads a separate archive can just unpack it in the main BoA directory).

**Comment by Username-N00b-is-not-available** [comment]
@Tormentor667 Here are updated versions of the addons:
[addons_behaviour.zip](https://github.com/Realm667/WolfenDoom/files/6343854/addons_behaviour.zip)
[addons_disable_hints.zip](https://github.com/Realm667/WolfenDoom/files/6343855/addons_disable_hints.zip)
[addons_efficiency.zip](https://github.com/Realm667/WolfenDoom/files/6343856/addons_efficiency.zip)
[addons_grenades_only.zip](https://github.com/Realm667/WolfenDoom/files/6343857/addons_grenades_only.zip)
[addons_instagib.zip](https://github.com/Realm667/WolfenDoom/files/6343858/addons_instagib.zip)
[addons_m2hb.zip](https://github.com/Realm667/WolfenDoom/files/6343859/addons_m2hb.zip)
[addons_melee.zip](https://github.com/Realm667/WolfenDoom/files/6343860/addons_melee.zip)
[addons_more_weapons.zip](https://github.com/Realm667/WolfenDoom/files/6343861/addons_more_weapons.zip)
[addons_movement.zip](https://github.com/Realm667/WolfenDoom/files/6343862/addons_movement.zip)
[addons_score.zip](https://github.com/Realm667/WolfenDoom/files/6343863/addons_score.zip)
[addons_vampire.zip](https://github.com/Realm667/WolfenDoom/files/6343864/addons_vampire.zip)

Missed explanation and screenshot for ```movement```:

movement
Ammo and large weapons now have weight and slow you down, you can drop unnecessary ones using the standard GZDoom weapon drop command. You can also roll by double-tapping a side movement key while crouching.

![movement](https://user-images.githubusercontent.com/59409713/115404072-a60f1d80-a1f5-11eb-9a0b-c611d1247a4e.png)

**Comment by Tormentor667** [comment]
Added, I guess we are done here for now :)

**Comment by DoomJedi** [comment]
DoomJuan offered a cool mutator idea - (as we have dices on BoA) you can play on casino tables with your gold, and win more gold or lose your gamble.

We can make moving dice and randomize its sides - and make engine "read" the actual roll result and compare to your gamble.
I can made dice sizes for all sides and we can animated those on a roll.

**Comment by Ozymandias81** [comment]
I guess you can also add bosses burning frames as a mutator then @DoomJedi , since those frames aren't exactly fitting modwise, but nice for a mutator I think.

**Comment by Tormentor667** [comment]
> I guess you can also add bosses burning frames as a mutator then @DoomJedi , since those frames aren't exactly fitting modwise, but nice for a mutator I think.

No actually not. As I said, the sprites clash totally with the solider burning sprites. I can’t accept this, even not as a mutator.
